src/rust/engine/graph/src/lib.rs

```diff
@@ -75,77 +99,144 @@ pub struct Entry<N: Node> {
+        waiters.push(send);
+        return recv
+          .map_err(|_| N::Error::invalidated())
+          .flatten()
+          .to_boxed();
+      }
+      &mut EntryState::Completed { ref result, .. } => {
+        return future::result(result.clone()).to_boxed();
+      }
+    };
 
-      self.state = Some(EntryState {
-        field: state.shared(),
-        start_time,
-      });
-      self.state(context, entry_id)
-    }
+    // swap in the new state and then recurse.
+    self.state = next_state;
+    self.get(context, entry_id)
   }
 
   ///
   /// If the Future for this Node has already completed, returns a clone of its result.
   ///
   fn peek(&self) -> Option<Result<N::Item, N::Error>> {
-    self
-      .state
-      .as_ref()
-      .and_then(|state| state.field.peek().map(unwrap_entry_res::<N>))
+    match &self.state {
+      &EntryState::Completed { ref result, .. } => Some(result.clone()),
+      _ => None,
+    }
   }
 
   ///
-  /// If the Node has started and has not yet completed, returns its runtime.
+  /// Called from the Executor when a Node completes.
   ///
-  fn current_running_duration(&self, now: Instant) -> Option<Duration> {
-    self.state.as_ref().and_then(|state| {
-      if state.field.peek().is_none() {
-        // Still running.
-        Some(now.duration_since(state.start_time))
-      } else {
-        None
+  fn complete(&mut self, result_run_token: RunToken, result: Result<N::Item, N::Error>) {
+    // Temporarily swap in `NotStarted` in order to take ownership of the state.
+    let state = mem::replace(&mut self.state, EntryState::initial());
+
+    // We care about exactly one case: a Running state with the same run_token. All other states
+    // represent various (legal) race conditions. See `RunToken`'s docs for more information.
+    self.state = match state {
+      EntryState::Running {
+        waiters,
+        start_time,
+        run_token,
+      } => {
+        if result_run_token == run_token {
+          // Notify all waiters (ignoring any that have gone away), and then store the value.
+          // A waiter will go away whenever they drop the `Future` `Receiver` of the value, perhaps
+          // due to failure of another Future in a `join` or `join_all`, or due to a timeout at the
+          // root of a request.
+          for waiter in waiters {
+            let _ = waiter.send(result.clone());
+          }
+          EntryState::Completed { result, run_token }
+        } else {
+          EntryState::Running {
+            waiters,
+            start_time,
+            run_token,
+          }
+        }
       }
-    })
+      s => s,
+    };
+  }
+
+  ///
+  /// If the Node has started and has not yet completed, returns its runtime.
+  ///
+  fn current_running_duration(&self, now: &Instant) -> Option<Duration> {
+    match &self.state {
+      &EntryState::Running { start_time, .. } => Some(now.duration_since(start_time)),
+      _ => None,
+    }
```

Comment:
This method is neat and is making me think about more instrumentation methods for the graph -- haven't thought of anything cool yet.